packages/menu/src/Menu.ts

```diff
@@ -201,30 +370,90 @@ export class Menu extends SpectrumElement {
         this.menuItems.forEach((item, i) => {
+            if (item.selected) {
+                selectedItemsMap.set(item, true);
+                selected.push(item.value);
+                selectedItems.push(item);
+            }
             if (i !== index) {
                 item.focused = false;
             }
         });
+        this.selectedItemsMap = selectedItemsMap;
+        this.selected = selected;
+        this.selectedItems = selectedItems;
         this.focusedItemIndex = index;
         this.focusInItemIndex = index;
     }
 
-    private prepItems = (): void => {
-        this.menuItems = [
-            ...this.querySelectorAll(`[role="${this.childRole}"]`),
-        ] as MenuItem[];
-        if (!this.menuItems || this.menuItems.length === 0) {
-            return;
+    // debounce update this.menuItems so they're in DOM order
+    private async updateMenuItems(): Promise<void> {
+        this.itemsChanged = true;
+
+        return new Promise((resolve) => {
+            // debounce so we only actually update once
+            requestAnimationFrame(() => {
+                if (!this.itemsChanged) {
+                    resolve();
+                }
+
+                this.menuItems = [];
+
+                const slotElements =
+                    this.menuSlot.assignedElements({ flatten: true }) || [];
+                for (const slotElement of slotElements) {
+                    const childItems: MenuItem[] =
+                        slotElement instanceof MenuItem
+                            ? [slotElement as MenuItem]
+                            : ([
+                                  ...slotElement.querySelectorAll(
+                                      `[role="${this.childRole}"]`
+                                  ),
+                              ] as MenuItem[]);
+                    for (const childItem of childItems) {
+                        if (this.menuItemSet.has(childItem)) {
+                            this.menuItems.push(childItem);
+                        }
+                    }
+                }
+
+                this.updateSelectedItemIndex();
+                this.updateItemFocus();
+                this.itemsChanged = false;
+                resolve();
+            });
+        });
+    }
+
+    private addItem(item: MenuItem) {
+        this.menuItemSet.add(item);
+        this.updateMenuItems();
+    }
+
+    private removeItem(itemToRemove: MenuItem) {
+        const newMenuItems: MenuItem[] = [];
+        for (const item of this.menuItems) {
+            if (item !== itemToRemove) {
+                newMenuItems.push(item);
+            }
         }
-        this.updateSelectedItemIndex();
+        this.menuItemSet.delete(itemToRemove);
+        this.updateMenuItems();
+    }
+
+    private async updateItemFocus() {
         const focusInItem = this.menuItems[this.focusInItemIndex] as MenuItem;
         if ((this.getRootNode() as Document).activeElement === this) {
```

Comment:
The work in `this.updateMenuItems()` _seems_ only necessary when we're about to do this work. Should we think about a second order of laziness beyond the batching/debouncing that you're doing now?

Reply:
I think we still need `this.menuItems` to be up to date (it's public, and you also should be able to programmatically update selected state).

Reply:
Makes sense.